### PR TITLE
feature: support configuring scan target for GlobalTransactionScanner

### LINKED ISSUE #2815: seata与springbatch发生出冲突
### Ⅰ. Issue Description
我在尝试与我们的项目整合时,启动报错,原因是seata代理与springbatch的@stepscope发生冲突
报错如下
### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'scopedTarget.batchMultiSendPartitioner': Scope 'step' is not active for the current thread; consider defining a scoped proxy for this bean if you intend to refer to it from a singleton; nested exception is java.lang.IllegalStateException: No context holder available for step scope
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:355)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	at org.springframework.aop.target.SimpleBeanTargetSource.getTarget(SimpleBeanTargetSource.java:35)
	at io.seata.spring.util.SpringProxyUtils.findTargetClass(SpringProxyUtils.java:57)
	at io.seata.spring.annotation.GlobalTransactionScanner.wrapIfNecessary(GlobalTransactionScanner.java:210)
	... 78 common frames omitted
```

### Ⅲ. Describe what you expected to happen
定位到springbatch的stepscope注解:
image
我猜测应该是与springbatch的@stepScope注解冲突导致此异常, 但我们的项目中多处使用到了springbatch 无法更换.
### Ⅴ. Anything else we need to know?
我确实不知道改如何解决此冲突,求助
### Ⅵ. Environment:

- JDK version :1.8
- Others: springcloud版本 E  seata版本 1.2.0(使用0.9.0也是这个错)

### PR DESCRIPTION
**Pull request by wangliang181230**
### Ⅰ. Describe what this PR did
feature: support custom scan target for GlobalTransactionScanner

### 配置方式1：配置只想扫描的package或需要排除的beanName：
```
seata:
  scan-packages: firstPackage,secondPackage
  excludes-for-scanning: firstBeanNameForExclude,secondBeanNameForExclude
```

### 配置方式2：实现扫描校验器接口，判断是否需要扫描：
扫描校验器接口：`io.seata.spring.annotation.ScannerChecker`
添加校验器的三种方式：
1. 创建`ScannerChecker`实现类的spring bean即可，seata会自动读取所有实现了该接口的bean.
2. 将实现类配置到各自模块下的`/META-INF/service/io.seata.spring.annotation.ScannerChecker`文件中。
3. 调用`GlobalTransactionScanner.addScannerCheckers(checkers)`静态方法添加。（注：上面两种方式最终也是由seata自行调用此方法添加的）

### seata提供的ScannerChecker
1. `io.seata.spring.annotation.scannercheckers.ScopeBeansScannerChecker`:
    目前可以排除 _添加了以下四种`@Scope`的bean_ : `request`, `session`, `job`, `step`.
    可自行调用`ScopeBeansScannerChecker.addExcludeScopes(scopeNames)`添加更多想要排除的scope类型。

2. `io.seata.spring.annotation.scannercheckers.ConfigBeansScannerChecker`:
    beanName为`*AutoConfiguration`和`*Properties`的bean排除掉。

### Ⅱ. Does this pull request fix one issue?

fixes #2815

### CODE REVIEW

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+public class ScopeBeansScannerExcluder implements ScannerExcluder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScopeBeansScannerExcluder.class);
+    private static final Set<String> EXCLUDE_SCOPE_SET = Collections.synchronizedSet(new HashSet<>());
+
+    static {
+        EXCLUDE_SCOPE_SET.add("request");
+        EXCLUDE_SCOPE_SET.add("session");
+        EXCLUDE_SCOPE_SET.add("step");
+        EXCLUDE_SCOPE_SET.add("job");
```
Comment (xingfudeshi) [inline spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java:53]: Please define constants for those names.

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            for (String scopeName : scopeNames) {
+                if (StringUtils.isNotBlank(scopeName)) {
+                    EXCLUDE_SCOPE_SET.add(scopeName.trim().toLowerCase());
+                }
+            }
+        }
+    }
+
+
+    public boolean isMatch(Object bean, String beanName, BeanDefinition beanDefinition) throws Throwable {
```
Comment (xingfudeshi) [inline spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java:72]: Missing '@Override'.

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ConfigBeansScannerExcluder.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Config scanner excluder.
+ *
+ * @author wang.liang
+ */
+@LoadLevel(name = "ConfigBeans", order = 100)
+public class ConfigBeansScannerExcluder implements ScannerExcluder {
+
+    public boolean isMatch(Object bean, String beanName, BeanDefinition beanDefinition) throws Throwable {
```
Comment (xingfudeshi) [inline spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ConfigBeansScannerExcluder.java:30]: Missing '@Override'.

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        if (bean instanceof ScopedProxyFactoryBean) {
+            return true; // exclude
```
Comment (xingfudeshi) [inline spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java:74]: Please don't write comments at the tail of line.Please check all places.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java [hunk 5]
```diff
@@ -60,17 +66,33 @@ public FailureHandler failureHandler() {
     @Bean
     @DependsOn({BEAN_NAME_SPRING_APPLICATION_CONTEXT_PROVIDER, BEAN_NAME_FAILURE_HANDLER})
     @ConditionalOnMissingBean(GlobalTransactionScanner.class)
-    public GlobalTransactionScanner globalTransactionScanner(SeataProperties seataProperties, FailureHandler failureHandler) {
+    public GlobalTransactionScanner globalTransactionScanner(
+            SeataProperties seataProperties, FailureHandler failureHandler,
```
Comment (xingfudeshi) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java:70]: SeataProperties is not supposed to be used here.It's only use for spring-boot auto-configurations.

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java (hunk 4 above)
Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java]: done

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ConfigBeansScannerExcluder.java (hunk 3 above)
Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ConfigBeansScannerExcluder.java:31]: done

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java (hunk 2 above)
Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java:77]: done

File: spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java (hunk 1 above)
Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/scannerexcluders/ScopeBeansScannerExcluder.java]: done

**Comment by wangliang1986** [comment]
> 
> 
> I left some comments.

OK.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java [hunk 6]
```diff
@@ -47,6 +47,10 @@
      * Whether use JDK proxy instead of CGLIB proxy
      */
     private boolean useJdkProxy = false;
+    /**
+     * Specifies which bean do not scanner in the GlobalTransactionScanner
```
Comment (xingfudeshi) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java:51]: Specifies beans that won't be scanned in the GlobalTransactionScanner

File: spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java [hunk 7]
```diff
@@ -200,9 +211,30 @@ protected Object wrapIfNecessary(Object bean, String beanName, Object cacheKey)
...
+                        BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
+                        for (ScannerExcluder excluder : SCANNER_EXCLUDER_SET) {
+                            try {
+                                if (excluder.isMatch(bean, beanName, beanDefinition)) {
+                                    return bean;
+                                }
+                            } catch (Throwable e) {
+                                String errorMsg = String.format("Do check need exclude failed: beanName=%s, excluder=%s",
+                                        beanName, excluder.getClass().getSimpleName());
+                                LOGGER.error(errorMsg, e);
```
Comment (xingfudeshi) [inline spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java:230]: Please use {}.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java (hunk 6 above)
Comment (wangliang1986) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java]: done.

File: spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java (hunk 7 above)
Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java:230]: There is no method `org.slf4j.Logger.error(String msg, Throwable t, Object... arguments);`. Only method `org.slf4j.Logger.error(String msg, Throwable t);` can be used.

Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java:230]: image

Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java:230]: Oh, it can be used like this. I've learned it. Thank you.

Comment (wangliang1986) [reply spring/src/main/java/io/seata/spring/annotation/GlobalTransactionScanner.java]: done

**Review by a364176773 — APPROVED** [review]
LGTM

**Pull request merged**